File: generated_code/GeminiPro/Java/081_SplayTree.java
/**
 * A self-contained Java class implementing a Splay Tree for an integer set.
 * Splay trees are self-balancing binary search trees with the additional property
 * that recently accessed elements are quick to access again.
 */
public class SplayTree {

    /**
     * Inner class representing a node in the Splay Tree.
     */
    private static class Node {
        int key;
        Node left;
        Node right;

        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * Performs a right rotation on the subtree rooted at node x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    /**
     * Performs a left rotation on the subtree rooted at node x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The main splay operation. It brings the node with the given key
     * (or the last accessed node if the key is not found) to the root.
     * @param root The root of the subtree to splay.
     * @param key The key to splay for.
     * @return The new root of the splayed subtree.
     */
    private Node splay(Node root, int key) {
        // Base case: root is null or key is present at root
        if (root == null || root.key == key) {
            return root;
        }

        // Key is in the left subtree
        if (key < root.key) {
            // Key is not in the tree, we are done
            if (root.left == null) {
                return root;
            }

            // Zig-Zig (Left Left)
            if (key < root.left.key) {
                // Recursively bring the key as root of left-left
                root.left.left = splay(root.left.left, key);
                // Do first rotation for root
                root = rightRotate(root);
            }
            // Zig-Zag (Left Right)
            else if (key > root.left.key) {
                // Recursively bring the key as root of left-right
                root.left.right = splay(root.left.right, key);
                // Do first rotation for root.left
                if (root.left.right != null) {
                    root.left = leftRotate(root.left);
                }
            }

            // Do second rotation for root. If root.left is null, it means the key was not found,
            // so we can't do the second rotation (the final Zig step).
            return (root.left == null) ? root : rightRotate(root);
        }
        // Key is in the right subtree
        else { // key > root.key
            // Key is not in the tree, we are done
            if (root.right == null) {
                return root;
            }

            // Zig-Zag (Right Left)
            if (key < root.right.key) {
                // Recursively bring the key as root of right-left
                root.right.left = splay(root.right.left, key);
                // Do first rotation for root.right
                if (root.right.left != null) {
                    root.right = rightRotate(root.right);
                }
            }
            // Zig-Zig (Right Right)
            else if (key > root.right.key) {
                // Recursively bring the key as root of right-right
                root.right.right = splay(root.right.right, key);
                // Do first rotation for root
                root = leftRotate(root);
            }

            // Do second rotation for root. If root.right is null, it means the key was not found,
            // so we can't do the second rotation (the final Zig step).
            return (root.right == null) ? root : leftRotate(root);
        }
    }

    /**
     * Searches for a key in the tree. If found, the corresponding node is
     * splayed to the root.
     *
     * @param key The key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        root = splay(root, key);
        if (root != null && root.key == key) {
            return root.key;
        }
        return null;
    }

    /**
     * Inserts a key into the tree. If the key already exists, the node
     * containing it is splayed to the root.
     *
     * @param key The key to insert.
     */
    public void insert(int key) {
        // Handle case of empty tree
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree for the key. This brings the closest node to the root.
        root = splay(root, key);

        // If the key is already present, we're done.
        if (root.key == key) {
            return;
        }

        // Otherwise, create a new node and make it the root, splitting the tree.
        Node newNode = new Node(key);
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Deletes a key from the tree.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the tree to bring the node to be deleted (or its parent) to the root.
        root = splay(root, key);

        // If the key is not in the tree, do nothing.
        if (root.key != key) {
            return;
        }

        // The node to be deleted is now at the root.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If there's no left child, the right subtree becomes the new tree.
            root = rightSubtree;
        } else {
            // Splay the left subtree to bring its largest element (the predecessor
            // of the deleted key) to its root.
            Node newRoot = splay(leftSubtree, key);

            // The new root (max of the left subtree) has no right child.
            // Attach the original right subtree there.
            newRoot.right = rightSubtree;
            root = newRoot;
        }
    }
}

File: generated_code/GeminiPro/Python/081_SplayTree.py
import sys

# It is recommended to increase recursion limit for deep trees,
# although the splaying operation keeps the tree relatively balanced.
# sys.setrecursionlimit(2000)

class SplayTree:
    """
    An implementation of a Splay Tree that acts as a dictionary-like set for integers.

    A splay tree is a self-adjusting binary search tree with the additional
    property that recently accessed elements are quick to access again.
    It performs basic operations such as insertion, look-up and removal
    in O(log n) amortized time.
    """

    class _Node:
        """A node in the Splay Tree."""
        def __init__(self, key, parent=None):
            self.key = key
            self.parent = parent
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on node x."""
        y = x.right
        if y:
            x.right = y.left
            if y.left:
                y.left.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        if y:
            y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on node x."""
        y = x.left
        if y:
            x.left = y.right
            if y.right:
                y.right.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
        if y:
            y.right = x
        x.parent = y

    def _splay(self, node):
        """
        Splays the given node to the root of the tree.
        This operation moves the node to the root through a series of rotations.
        """
        while node.parent:
            parent = node.parent
            grandparent = parent.parent
            if not grandparent:
                # Zig case
                if node == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif node == parent.left and parent == grandparent.left:
                # Zig-Zig case (left-left)
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif node == parent.right and parent == grandparent.right:
                # Zig-Zig case (right-right)
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            elif node == parent.right and parent == grandparent.left:
                # Zig-Zag case (left-right)
                self._left_rotate(parent)
                self._right_rotate(grandparent)
            else:
                # Zig-Zag case (right-left)
                self._right_rotate(parent)
                self._left_rotate(grandparent)

    def search(self, key):
        """
        Searches for a key in the tree.

        If the key is found, the node containing the key is splayed to the root.
        If the key is not found, the last accessed node (the parent of the
        hypothetical key) is splayed to the root.

        Args:
            key: The integer key to search for.

        Returns:
            True if the key is found, False otherwise.
        """
        if not self.root:
            return False

        current = self.root
        last_visited = None
        while current:
            last_visited = current
            if key == current.key:
                self._splay(current)
                return True
            elif key < current.key:
                current = current.left
            else:
                current = current.right
        
        # If key not found, splay the last visited node
        if last_visited:
            self._splay(last_visited)
            
        return False

    def insert(self, key):
        """
        Inserts a key into the tree.

        If the key already exists, the existing node is splayed to the root.
        If the key is new, it is inserted and the new node is splayed to the root.

        Args:
            key: The integer key to insert.
        """
        if not self.root:
            self.root = self._Node(key)
            return

        current = self.root
        parent = None
        while current:
            parent = current
            if key == current.key:
                # Key already exists, splay it and we are done
                self._splay(current)
                return
            elif key < current.key:
                current = current.left
            else:
                current = current.right

        new_node = self._Node(key, parent)
        if key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node
        
        self._splay(new_node)

    def delete(self, key):
        """
        Deletes a key from the tree.

        The node to be deleted (or its parent if not found) is first splayed
        to the root. If the key is found at the root, it is removed, and the
        two resulting subtrees are merged.

        Args:
            key: The integer key to delete.
        """
        if not self.root:
            return

        # Splay the node (or its would-be parent) to the root
        self.search(key)

        # If the key is not at the root after splaying, it doesn't exist
        if self.root.key != key:
            return

        # Now the node to delete is the root
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # No left child, the right subtree becomes the new tree
            self.root = right_subtree
            if right_subtree:
                right_subtree.parent = None
        else:
            # Disconnect left subtree from the old root
            left_subtree.parent = None
            
            # Find the maximum element in the left subtree
            max_node = left_subtree
            while max_node.right:
                max_node = max_node.right
            
            # Splay this maximum element to the root of the left subtree
            # After this splay, the new root of the left subtree has no right child
            self._splay(max_node)
            
            # Attach the original right subtree as the right child of the new root
            max_node.right = right_subtree
            if right_subtree:
                right_subtree.parent = max_node
            
            self.root = max_node

    def get_inorder_keys(self):
        """Helper method to get keys in-order for testing/visualization."""
        if not self.root:
            return []
        keys = []
        def _inorder_traverse(node):
            if node:
                _inorder_traverse(node.left)
                keys.append(node.key)
                _inorder_traverse(node.right)
        _inorder_traverse(self.root)
        return keys

# Example Usage (optional, for demonstration)
if __name__ == '__main__':
    st = SplayTree()
    
    # Insert elements
    keys_to_insert = [10, 20, 30, 40, 50, 25]
    print(f"Inserting keys: {keys_to_insert}")
    for k in keys_to_insert:
        st.insert(k)
        print(f"After inserting {k}, root is: {st.root.key}")
        print(f"  In-order traversal: {st.get_inorder_keys()}")
    
    print("\n" + "="*30 + "\n")
    
    # Search for an element
    key_to_search = 30
    print(f"Searching for {key_to_search}...")
    found = st.search(key_to_search)
    print(f"Found {key_to_search}: {found}")
    if st.root:
        print(f"After searching for {key_to_search}, root is: {st.root.key}")
    
    print("\n" + "="*30 + "\n")

    # Search for another element
    key_to_search = 10
    print(f"Searching for {key_to_search}...")
    found = st.search(key_to_search)
    print(f"Found {key_to_search}: {found}")
    if st.root:
        print(f"After searching for {key_to_search}, root is: {st.root.key}")

    print("\n" + "="*30 + "\n")

    # Search for a non-existent element
    key_to_search = 99
    print(f"Searching for non-existent key {key_to_search}...")
    found = st.search(key_to_search)
    print(f"Found {key_to_search}: {found}")
    if st.root:
        print(f"After searching for {key_to_search}, root is: {st.root.key} (last accessed node)")

    print("\n" + "="*30 + "\n")

    # Delete an element
    key_to_delete = 30
    print(f"Deleting {key_to_delete}...")
    st.delete(key_to_delete)
    if st.root:
        print(f"After deleting {key_to_delete}, root is: {st.root.key}")
        print(f"  In-order traversal: {st.get_inorder_keys()}")
    
    print("\n" + "="*30 + "\n")

    # Delete another element
    key_to_delete = 50
    print(f"Deleting {key_to_delete}...")
    st.delete(key_to_delete)
    if st.root:
        print(f"After deleting {key_to_delete}, root is: {st.root.key}")
        print(f"  In-order traversal: {st.get_inorder_keys()}")